### PR TITLE
Add head reduction flags (`lazy head beta` etc)

### PR DESCRIPTION
**Pull request by SkySkimmer**
NB: the `cbv` code doesn't seem to have a weak mode (I guess it would
just mean not reducing under binders).

Overlays:
- https://github.com/ejgallego/coq-serapi/pull/360

### CODE REVIEW

**Comment by SkySkimmer** [comment]
@coqbot run full ci

**Review by jfehrle — COMMENTED** [review]
Some suggestions.  Would be good to add the index entries.

**Review by JasonGross — APPROVED** [review]
This is great, thank you!

I would also be quite interested in modifiers that meant "don't go underneath lambdas".  Should I open a separate issue for that?

File: test-suite/output/reduction.v [hunk 1]
```diff
@@ -11,3 +11,6 @@ Eval simpl in (fix plus (n m : nat) {struct n} : nat :=
 
 Eval hnf in match (plus (S n) O) with S n => n | _ => O end.
 
+Eval weak simpl in 2 + 2.
```
Comment (JasonGross) [inline test-suite/output/reduction.v:14]: Please also test `2 + (2 + 2)` (behavior on terms with non-normal arguments`) and `hold (fun x => 1 + x)` for `Axiom hold : forall {T}, T.` (when the arguments to the whnf have lambdas)

**Comment by SkySkimmer** [comment]
If we want to add modifiers we should probably not go with this syntax and instead put the modifiers among the flags (beta etc)
The situation for simpl with flags is kinda strange btw, it ignores the flags unless simpliscbn AFAICT

**Comment by JasonGross** [comment]
> If we want to add modifiers we should probably not go with this syntax and instead put the modifiers among the flags (beta etc)

I'd be fine with `weak` meaning the thing in this PR for `lazy`, `cbn`, and `simpl`, and meaning "don't go under binders" for `cbv`, `vm_compute` and `native_compute`.

**Comment by jfehrle** [comment]
> I'd be fine with `weak` meaning the thing in this PR for `lazy`, `cbn`, and `simpl`, and meaning "don't go under binders" for `cbv`, `vm_compute` and `native_compute`.

Only if we document the behavior/differences!

File: doc/sphinx/proofs/writing-proofs/equality.rst [hunk 2]
```diff
@@ -485,7 +485,7 @@ which reduction engine to use.  See :ref:`type-cast`.)  For example:
 
       Abort.
 
-.. tacn:: lazy {? @reductions } @simple_occurrences
+.. tacn:: {? weak } lazy {? @reductions } @simple_occurrences
```
[the remark this answers is not in the document]
Comment (Zimmi48) [reply doc/sphinx/proofs/writing-proofs/equality.rst:488]: Even with your changes from #17507, this won't work, because the names should come after the signature, and should include `cbv` (in the middle, with the current implementation of #17507).

**Comment by SkySkimmer** [comment]
>I'd be fine with weak meaning the thing in this PR for lazy, cbn, and simpl, and meaning "don't go under binders" for cbv, vm_compute and native_compute.

But "go in subterms but not under binders" is also meaningful for cbn & co.

File: doc/sphinx/proofs/writing-proofs/equality.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (jfehrle) [reply doc/sphinx/proofs/writing-proofs/equality.rst:488]: Answers are $5.  Correct answers are $10.  Dumb looks are still free :-)  Please DWIM.

**Comment by JasonGross** [comment]
> > I'd be fine with weak meaning the thing in this PR for lazy, cbn, and simpl, and meaning "don't go under binders" for cbv, vm_compute and native_compute.
> 
> But "go in subterms but not under binders" is also meaningful for cbn & co.

That's true.  But it's also the natural meaning of "weak" for cbv ("stop once you're guaranteed that there can be no more redexs in the head position").  But in any case, I care more about having the functionality at all than about what the syntax is.  And I support merging this PR with the current syntax.  (Though I do think it'd be cleaner overall to have two flags instead of prefixes, one for "stop when there are guaranteed to be no redexs in the head position", and another for "do not traverse into the bodies of lambdas/Pis".)

**Comment by SkySkimmer** [comment]
`weak` was the wrong term, it should have been `head`. Now changed.
Also changed to use the regular postfix flag system instead of a special prefix.

**Comment by herbelin** [comment]
It is a matter of definition, but in traditional lambda-calculus semantics, head includes going under lambdas, while weak-head goes only under applications (i.e. morally to reach either a form Evalcontext[x] blocking on a term variable or a form Value, blocking on a missing context of evaluation). I don't think that weak is used alone, but if it were, I would probably understand it as weak-head.

**Comment by JasonGross** [comment]
> Btw, I'm a bit confused as I thought the original user request #13881 was about (3), not (2), wasn't it?

I think I'm confused about the difference between weak head and strong head.  Is the difference just whether or not delta is included?  I think I meant to ask about strong head, though, not weak head (`hnf` is strong head, right?).  And I'm actually more interested in the fourth ML-like strategy that doesn't recurse under binders, though I think having hnf-like behavior for the other reduction engines is nice

**Comment by gares** [comment]
`fun x => ....` is in weak head normal form, no matter what `...` is

**Comment by maximedenes** [comment]
In my understanding, "weak head" does not go through binders, whereas "strong head" does. I.e. `fun x => (fun y => y) x` is a normal form for the first, but reduces to `fun x => x` for the second.

The ML-like strategy is also weak, but reduces in arguments, not just in the head position (see the example in my previous comment).

**Comment by SkySkimmer** [comment]
So on `(fun x y => x) ((fun z => z) (1 + 0, fun _ => 1 + 0))`:
- `fun y => (1, fun _ => 1)` is the result of full reduction (regardless of strategy)
- `fun y => ((fun z => z) (1 + 0, fun _ => 1 + 0))` is the cbn weak head result and also the weak non-head result (weak: don't recurse under `fun y`)
- `fun y => (1 + 0, fun _ => 1 + 0)` is the cbn strong head result (head: don't recurse in subterms of `(_, _)`)
- `fun y => (1, fun _ => 1 + 0)` is the cbv weak result (cbv doesn't distinguish between head and non-head)

On `Some ((fun x y => x) (1 + 0))`
- `Some (fun y => 1)` is the result of full reduction and cbv weak 
- `Some ((fun x y => x) (1 + 0))` is the result of cbn weak and strong head 
- `Some (fun y => 1 + 0)` is the result of cbn weak non-head 

(full reduction = cbn strong non-head = cbv strong)

Depending on if the head reduces a lambda or not, cbn weak head has the same result as either cbn weak non-head (when it's a lambda) or cbn strong head (when it's not a lambda)

Are the above also how other people understand "weak" "strong" and "head"?

--------------

vm_compute and native_compute are cbv strong with no support for flags (delta beta etc)

simpl is full reduction with some weird delta and refolding behaviour (?)

hnf is weak head cbn with some weird delta and refolding behaviour (not quite the same as simpl) followed by full betaiota (?)
For instance
~~~coq
Eval hnf in (fun x (y:nat) => x) (let z := (1 + 0, fun _:nat => 1 + 0) in z).
(*  fun _ : nat => let z := (1 + 0, fun _ : nat => 1 + 0) in z *)
Eval hnf in (fun x => Some x) (let z := (1 + 0, fun _:nat => 1 + 0) in z).
(*  Some (let z := (1 + 0, fun _ : nat => 1 + 0) in z) *)
~~~

lazy is cbn with sharing (?)

**Comment by herbelin** [comment]
I can clarify my view (which I believe coincides with the usage):
- the good way to understand weak is to reason in a polarized lambda-calculus (typically a linear variant of call-by-push-value, CBPV)
- in such calculus, there are different forms of cbn and different forms of cbv, while cbneed is even more complicated
  - in such calculus, a cbn application has the form t(↓u) meaning that the evaluation of the argument is delayed
  - in such calculus, a cbv application can typically be left-to-right, right-to-left, n-ary right-to-left:
    - in Coq, it seems that it is an n-ary right-to-left application, that is that "t u1...un" is "let (x1...xn):=(u1...un) in let ↓f:=t in fx1..xn", meaning that the arguments of an application are first evaluated in parallel, then the function is evaluated, then the beta-redex, if any, is evaluated
- weak does not specifically refer to going under binders or not but about going under a shift of polarity (knowing that a lambda includes a shift of polarity)
- there are several ways to define evaluation of connectives depending on where we put a shift, whose role is to delay evaluation:
  - for instance, we can consider each natural number as one value (corresponding to the encoding  μX⁺.(1 ⊕ X⁺) in CBPV) or 0 and S as constructors of values with delays (corresponding to the encoding μX⁺.(1 ⊕ ↓↑X⁺) in CBPV, or, similarly μX⁻.↑(1 ⊕ ↓X⁻))
  - traditionally, cbn and cbneed are considered to have datatypes which evaluate lazily (one constructor after the other) while cbv is considered to have datatypes which evaluate eagerly (i.e. all constructors glued together, unless there is a function type in the constructor type); afaicj, this is also what weak-head does in Coq for cbn, cbneed and cbv
- the fact that finding the head of an application in lambda-calculus requires to go under binary application nodes is (in my opinion) a defect of natural deduction; evaluation has better properties when presented in sequent calculus, that is, using an abstract machine; in particular, the good notions are the notions of evaluation contexts E and values V, in such a way that a weak-head normal form is either of the form E[x], i.e. the state <x|E> of an abstract machine, or V, i.e. the state <V|end> of an abstract machine

If we want to do things in an as principled way as possible, we may take some time to do it. In particular, I doubt that cbn and cbv are currently implemented in a way which we know how to document clearly. If you want, I may try within a week to exactly clarify what the cbn and cbv semantics are currently doing. Or we could also do it in a collaborative way, by giving the exact definition of evaluation contexts, values and terms on which Coq's cbn and cbv rely, what would characterize exactly where delays (i.e. shifts) are inserted. Moreover, while it is rather clear when beta, iota, fix, cofix have to apply (they are all generalizations of beta), it is certainly worth clarifying when zeta is supposed to be applied. For instance, what is supposed to be the weak reduction of `(let _ := y z in fun x => x) 0` in cbv: to block on `y` in the absence of zeta but to return 0 in the presence of zeta?

**Comment by jfehrle** [comment]
> So on `(fun x y => x) ((fun z => z) (1 + 0, fun _ => 1 + 0))`:

This would be good to add to the documentation somewhere.

**Comment by JasonGross** [comment]
What is the state of this PR?

**Comment by JasonGross** [comment]
What was blocking this PR?  I would quite like to see something like this

**Comment by herbelin** [comment]
I think this PR can and should go in.

(We could eventually add a weak mode to cbv too...)

File: plugins/ssr/ssrcommon.ml [hunk 3]
```diff
@@ -1192,9 +1192,7 @@ let rec decompose_assum env sigma orig_goal =
 
 let tclFULL_BETAIOTA = Goal.enter begin fun gl ->
   let r, _ = Redexpr.reduction_of_red_expr (Goal.env gl)
-    Genredexpr.(Lazy {
-      rBeta=true; rMatch=true; rFix=true; rCofix=true;
-      rZeta=false; rDelta=false; rConst=[]}) in
+    Genredexpr.(Lazy Redops.all_flags) in
```
Comment (rlepigre) [inline plugins/ssr/ssrcommon.ml:1195]: Is this change intended? Both `rZeta` and `rDelta` used to be `false`, but are now `true`.

Comment (SkySkimmer) [reply plugins/ssr/ssrcommon.ml:1195]: not intended

**Comment by SkySkimmer** [comment]
@coqbot run full ci

**Comment by SkySkimmer** [comment]
@coqbot run full ci

**Comment by SkySkimmer** [comment]
IDK why test-suite:base https://gitlab.inria.fr/coq/coq/-/jobs/3440686 failed, it timeouted but testing locally that line takes below detectable time.

**Comment by SkySkimmer** [comment]
@coqbot run full ci

File: tactics/redexpr.ml [hunk 4]
```diff
@@ -209,18 +208,14 @@ let out_with_occurrences (occs,c) =
 
 let e_red f env evm c = evm, f env evm c
 
-let head_style = false (* Turn to true to have a semantics where simpl
-   only reduce at the head when an evaluable reference is given, e.g.
-   2+n would just reduce to S(1+n) instead of S(S(n)) *)
-
-let contextualize f g = function
+let contextuallize f = function
```
Comment (yannl35133) [inline tactics/redexpr.ml:211]: Trivial question but why the double 'l' ? AFAICT this variant doesn't appear in dictionaries

Comment (SkySkimmer) [reply tactics/redexpr.ml:211]: I have no idea how this change appeared

**Comment by herbelin** [comment]
To support weak CBV, I made #18190. If OK for everyone, I would advocate to merge it first so that the current PR is complete (the current PR would just have to change its line "Cbv (Head, _) -> ..." in redexpr.ml).

**Comment by SkySkimmer** [comment]
I would rather merge this PR first so we don't have dead code

**Comment by herbelin** [comment]
OK if you insist. Then how to proceed? You fix the `ll` of contextualize and update the serapi orverlay, then I merge?

**Review by herbelin — APPROVED** [review]
OK to me.

Note that the following will later have to be done:
- weak cbv (but #18190 will do)
- support for the `head` flag at the syntax level in Ltac2

Regarding terminology: `head` looks good because more intuitive than `weak-head` (e.g. using the word `whd` would have looked a bit jargon).

**Comment by herbelin** [comment]
Also: will it be "acceptable practice" that the weak cbv PR modifies the change log of this PR?

File: tactics/redexpr.ml (hunk 4 above)
Comment (SkySkimmer) [reply tactics/redexpr.ml:211]: changed

**Comment by JasonGross** [comment]
I expect the minimization will not go anywhere without help; the file takes 10s to run and is 3700loc...

**Comment by SkySkimmer** [comment]
Looks like it was a temporary problem anyway, probably some incompatible mathcomp change.

**Comment by SkySkimmer** [comment]
@herbelin this should be ready to merge

**Comment by herbelin** [comment]
@coqbot: merge now (finally!)
